### PR TITLE
R4R: Refactor distribution query using Querier

### PR DESCRIPTION
**Pull request by chengwenxi**
Ref: https://github.com/irisnet/irishub/issues/1374

### CODE REVIEW

File: modules/distribution/keeper/querier.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+func queryDelegatorWithdrawAddress(ctx sdk.Context, _ []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params QueryDelegatorParams
+	err := k.cdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	// cache-wrap context as to not persist state changes during querying
+	ctx, _ = ctx.CacheContext()
```
Comment (HaoyangLiu) [inline modules/distribution/keeper/querier.go]: There is no need for cache.

File: modules/distribution/keeper/querier.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+// nolint
+const (
+	QueryWithdrawAddr          = "withdraw_addr"
+	QueryDelegationDistInfo    = "delegation_dist_info"
+	QueryAllDelegationDistInfo = "all_delegation_dist_info"
+	QueryValidatorDistInfo     = "validator_dist_info"
```
Comment (HaoyangLiu) [inline modules/distribution/keeper/querier.go]:
```
	QueryDelegationDistInfo    = "delegation_dist_info"
	QueryAllDelegationDistInfo = "all_delegation_dist_info"
	QueryValidatorDistInfo     = "validator_dist_info"
```
The above three querier doesn't make too much sense. It is not necessary to implement them

Comment (chengwenxi) [reply modules/distribution/keeper/querier.go]: Will decide which APIs are supported in the future, but not now.

File: modules/distribution/keeper/querier.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+		if vdi.OperatorAddr.Equals(selfValidator.GetOperator()) {
+			selfVdi = vdi
+		}
+		feePool = newFeePool
+		return false
+	}
+	k.stakeKeeper.IterateDelegations(ctx, params.Address, operationAtDelegation)
+
+	// get all validator rewards
+	if selfVdi.OperatorAddr.Equals(sdk.ValAddress(params.Address)) {
```
Comment (HaoyangLiu) [inline modules/distribution/keeper/querier.go]: If the delegator is not the operator of any validators, selfVdi may be a empty struct. `selfVdi.OperatorAddr` may be nil. Shall we judge whether `selfVdi.OperatorAddr` is empty or not?

File: modules/distribution/keeper/querier.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+	return QueryRewardsParams{
+		Address: address,
+	}
+}
+
+type Rewards struct {
+	Total       sdk.Coins           `json:"total"`
+	Delegations []DelegationsReward `json:"delegations"`
+	Commission  sdk.Coins           `json:"commission"`
+}
```
Comment (HaoyangLiu) [inline modules/distribution/keeper/querier.go]: Due to decimal truncation, `total` may not equal to the sum of delegations reward and commission. Is this acceptable?

File: modules/distribution/keeper/querier.go (hunk 3 above)
Comment (chengwenxi) [reply modules/distribution/keeper/querier.go]: The `address. Equals` function Has checked if it is empty.

File: modules/distribution/keeper/querier.go (hunk 4 above)
Comment (chengwenxi) [reply modules/distribution/keeper/querier.go]: I think no problem, any good suggestions?

File: modules/distribution/keeper/querier.go (hunk 3 above)
Comment (chengwenxi) [reply modules/distribution/keeper/querier.go]: But `selfVdi` may be nil, should add check `selfVdi == nil` .

File: modules/distribution/keeper/querier.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+		validator := k.stakeKeeper.Validator(ctx, del.GetValidatorAddr())
+		vdi := k.GetValidatorDistInfo(ctx, del.GetValidatorAddr())
+		wc := k.GetWithdrawContext(ctx, del.GetValidatorAddr())
+		wc.FeePool = feePool
+		distInfo := k.GetDelegationDistInfo(ctx, del.GetDelegatorAddr(), del.GetValidatorAddr())
+		_, vdi, newFeePool, diWithdraw := distInfo.WithdrawRewards(log.NewNopLogger(), wc, vdi, validator.GetDelegatorShares(), del.GetShares())
+		totalWithdraw = totalWithdraw.Plus(diWithdraw)
+		rewardTruncate, _ := diWithdraw.TruncateDecimal()
+		rewards.Delegations = append(rewards.Delegations, DelegationsReward{del.GetValidatorAddr(), rewardTruncate})
+		if vdi.OperatorAddr.Equals(selfValidator.GetOperator()) {
```
Comment (chengwenxi) [inline modules/distribution/keeper/querier.go]: Check selfValidator is `nil`

**Pull request merged**